runtime/buffer.h

```diff
@@ -979,6 +1012,63 @@ SLINKY_NO_STACK_PROTECTOR void for_each_tile(span<const index_t> tile, const raw
 // Value for use in tile tuples indicating the dimension should be passed unmodified.
 static constexpr index_t all = std::numeric_limits<index_t>::max();
 
+namespace internal {
+
+inline void increment(std::size_t rank, index_t* index, const index_t* min, const index_t* max) {
+  index_t& i = *index;
+  ++i;
+  if (i > *max && rank > 1) {
+    i = *min;
+    increment(rank - 1, index + 1, min + 1, max + 1);
+  }
+}
+
+template <typename It>
+class iterator_range {
+  It begin_, end_;
+
+public:
+  iterator_range(It begin, It end) : begin_(std::move(begin)), end_(std::move(end)) {}
+
+  const It& begin() const { return begin_; }
+  const It& end() const { return end_; }
+};
+
+class index_iterator {
+  std::vector<index_t> index;
+  std::vector<index_t> min;
+  std::vector<index_t> max;
+
+public:
+  index_iterator() {}
+  index_iterator(std::vector<index_t> index, std::vector<index_t> min, std::vector<index_t> max)
+      : index(std::move(index)), min(std::move(min)), max(std::move(max)) {}
+
+  bool operator==(const index_iterator& r) const {
+    return index.size() == r.index.size() && internal::equal_small_n(index.data(), r.index.data(), index.size());
+  }
+  bool operator!=(const index_iterator& r) const { return !operator==(r); }
+
+  const std::vector<index_t>& operator*() const { return index; }
+
+  index_iterator operator++(int) {
+    index_iterator result(*this);
+    ++*this;
+    return result;
+  }
+  index_iterator& operator++() {
+    internal::increment(index.size(), index.data(), min.data(), max.data());
+    return *this;
+  }
+};
+
+}  // namespace internal
+
+// Return an iterator_range that iterators indices of a buffer, starting at dimension `min_dim`. This is not an
```

Comment:
What does it mean to start at `min_dim`? if min_dim == 0 iterate over each element, if min_dim == 1 iterate over each row, etc?